File: heeheej/week05/PRG_1844.py
# 게임 맵 최단거리
# 백준 환경에서는 잘 안헷갈리는데 x축 y축을 혼동해버렸다. 왜지? 자알 생각하고 풀자.

dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]
from collections import deque

def solution(maps):
    N, M = len(maps), len(maps[0])  # 이거 헷갈리면 안돼...
    queue = deque()
    queue.append((0, 0))
    visited = [[-1] * M for _ in range(N)]  # 얘도....
    visited[0][0] = 1

    while queue:
        x, y = queue.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if nx < 0 or nx >= N or ny < 0 or ny >= M or visited[nx][ny] != -1 or maps[nx][ny] == 0:
                continue
            if nx == N - 1 and ny == M - 1:
                return visited[x][y] + 1
            visited[nx][ny] = visited[x][y] + 1
            queue.append((nx, ny))
    return -1

File: teamzamong/week05/PRG_1844.java
import java.util.*;

class PRG_1844 {

  static int[] dr = { -1, 1, 0, 0 };
  static int[] dc = { 0, 0, -1, 1 };

  public int solution(int[][] maps) {
    int n = maps.length;
    int m = maps[0].length;

    boolean[][] visited = new boolean[n][m];
    Queue<int[]> queue = new ArrayDeque<>();
    queue.offer(new int[] { 0, 0 });

    int count = 0;
    while (!queue.isEmpty()) {

      int size = queue.size();
      for (int i = 0; i < size; ++i) {
        int[] curr = queue.poll();

        if (curr[0] == n - 1 && curr[1] == m - 1)
          return count + 1;

        for (int d = 0; d < 4; ++d) {
          int nr = curr[0] + dr[d];
          int nc = curr[1] + dc[d];

          if (!inRange(nr, nc, n, m) || maps[nr][nc] == 0 || visited[nr][nc])
            continue;

          visited[nr][nc] = true;
          queue.offer(new int[] { nr, nc });
        }
      }
      ++count;
    }

    return -1;
  }

  static boolean inRange(int r, int c, int n, int m) {
    return r >= 0 && c >= 0 && r < n && c < m;
  }
}